### PR TITLE
Adding new algorithm for Longest Increasing Subsequence.

### LINKED ISSUE #153: Longest Increasing Subsequence
I would like to do an implementation of Longest Increasing Subsequence in C++.
Problem statement:
Given an unsorted array of integers, return the longest increasing subsequence.

### PR DESCRIPTION
**Pull request by hijajoo**
Adding algorithm and testcases for Longest Increasing Subsequence in folder dynamic_programming.

- [X] Have read the [contribution guidelines][contrib-guidelines]
- [X] Closes an issue: close #153  
- [X] Includes unit tests (or unit tests have already been added)

[contrib-guidelines]: https://github.com/faheel/Algos/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: C++/source/dynamic_programming/longest_increasing_subsequence.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    /*
+       Reverse the vector ans so that it has values in ascending order
+     */
+    reverse(ans.begin(), ans.end());
+    return ans;
+}
+
+#ifndef LONGEST_INCREASING_SUBSEQUENCE_TEST
+int main() {
+    return 0;
```
Comment (faheel) [inline C++/source/dynamic_programming/longest_increasing_subsequence.cpp]: An empty `main()` is not a good idea. Can you please add some code to take input from the user, call `longest_increasing_subsequence()`, and display it's result too?

File: C++/test/dynamic_programming/longest_increasing_subsequence.test.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    SECTION("Size is as expected") {
+        REQUIRE(lis.size() == expected_lis.size());
+    }
+    for(int i = 0; i < (int)lis.size(); i++) {
+        SECTION("Checking value at index " + i) {
+            REQUIRE(lis[i] == expected_lis[i]);
+        }
+    }
+}
+
```
Comment (faheel) [inline C++/test/dynamic_programming/longest_increasing_subsequence.test.cpp]: Please add 2 more test cases, one in which some numbers are negative, and one in which all numbers are negative in the expected longest increasing subsequence.

File: C++/source/dynamic_programming/longest_increasing_subsequence.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            max_i = i;
+        }
+    }
+    return max_i;
+}
+
+vector<int> longest_increasing_subsequence(const vector<int> &a) {
+    vector<int> lis_count(a.size());
+    vector<int> lis_prev(a.size());
+    vector<int> ans;
```
Comment (faheel) [inline C++/source/dynamic_programming/longest_increasing_subsequence.cpp]: Declare variables just before they're used for the first time (which would be line 80 in case of `ans`), and not at the very top of a function.

File: C++/source/dynamic_programming/longest_increasing_subsequence.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+int find_prev_smaller(const vector<int> &a, int x) {
+    for(int i = x-1; i >= 0; i--) {
+        if(a[i] < a[x]) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+int index_of_max(const vector<int> &a) {
```
Comment (faheel) [inline C++/source/dynamic_programming/longest_increasing_subsequence.cpp]: Add a comment that briefly describes what this function does.
Also, consider a better name for `a` (`values`? `input`?).

File: C++/source/dynamic_programming/longest_increasing_subsequence.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+ */
+
+#include <iostream>
+#include <vector>
+#include <climits>
+#include <algorithm>
+
+using namespace std;
+
+int find_prev_smaller(const vector<int> &a, int x) {
```
Comment (faheel) [inline C++/source/dynamic_programming/longest_increasing_subsequence.cpp]: Add a comment that briefly describes what this function does.
Also, consider a better name for `a` (`values`? `input`?) and `x` (`current_index`?).

**Review by faheel — CHANGES_REQUESTED** [review]
Thank you for contributing! There are a few changes that I think would be required before merging the PR.

File: C++/source/dynamic_programming/longest_increasing_subsequence.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
+/*
+   Longest Increasing Subsequence
+   ------------------------------
+   Given an unsorted array of integers, return the longest increasing subsequence
+
+   Time complexity
+   --------------
+   O(N), where N is size of array
```
Comment (faheel) [inline C++/source/dynamic_programming/longest_increasing_subsequence.cpp]: I think it's *O*(N<sup>2</sup>) since for every index you call `find_previous_smaller()`, which takes linear time. So for `N` number of indices, it would take *O*(N<sup>2</sup>) time.

**Comment by hijajoo** [comment]
Travis CI errors seem to be unrelated to my changes.

File: C++/test/dynamic_programming/longest_increasing_subsequence.test.cpp [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+TEST_CASE( "Vector has single element" , "[longest_increasing_subsequence]" ) {
+    vector<int> a = {5};
+    vector<int> expected_lis = {5};
+    vector<int> lis = longest_increasing_subsequence(a);
+    SECTION("Size is as expected") {
+        REQUIRE(lis.size() == expected_lis.size());
+    }
+    for(int i = 0; i < (int)lis.size(); i++) {
+        SECTION("Checking value at index " + i) {
```
Comment (faheel) [inline C++/test/dynamic_programming/longest_increasing_subsequence.test.cpp:14]: `SECTION` block is not required here (and in other test cases too).

**Review by faheel — CHANGES_REQUESTED** [review]

**Comment by faheel** [comment]
@hijajoo I found a test case that gives wrong output:
Input: `2 6 4 8 7 -7 5 2 1 4 6 8 9 0 2 3` (size 16)
Expected output: `-7 1 4 6 8 9`
Actual output: `-7 1 4 6 8 9 0 2 3`

**Comment by alxmjo** [comment]
Looks like this hasn't been updated in a while. Closing.
